modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/AssignmentsTracker.java

```diff
@@ -241,4 +241,87 @@ private static String collectKeysFromEventAsString(WatchEvent event) {
                 .collect(Collectors.joining(","));
     }
 
+    /**
+     * Prepares assignments for logging using the following structure:
+     * consistentId=[peers=[1_part_1, 1_part_2], learners=[2_part_0]].
+     *
+     * @param assignmentsMap assignments to be logged.
+     * @return String representation of assignments.
+     */
+    private static String prepareAssignmentsForLogging(Map<ReplicationGroupId, TokenizedAssignments> assignmentsMap) {
+        class NodeAssignments {
+            private List<ReplicationGroupId> peers;
+            private List<ReplicationGroupId> learners;
+
+            private NodeAssignments() {
+            }
+
+            private void addReplicationGroupId(ReplicationGroupId replicationGroupId, boolean isPeer) {
+                List<ReplicationGroupId> peersOrLearners;
+
+                if (isPeer) {
+                    if (peers == null) {
```

Comment:
We may initialize the fields with empty lists (I bet Linked lists are fine there in terms of memory) and emptiness check right through `list.isEmpty()` only below.

Reply:
Yep, it was my explicit intention not to allocate objects until we know that the collection is not empty, to avoid any GC pressure. IMHO, this does not in any way impair the readability of the code.